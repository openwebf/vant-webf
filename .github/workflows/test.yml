name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - dev

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: npm test

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v1
      with:
        token: ${{ secrets.CODECOV_TOKEN }}

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
