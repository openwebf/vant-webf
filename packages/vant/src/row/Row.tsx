import {
  computed,
  defineComponent,
  type PropType,
  type ComputedRef,
  type InjectionKey,
  type ExtractPropTypes,
} from 'vue';
import {
  truthProp,
  makeStringProp,
  makeNumericProp,
  createNamespace,
} from '../utils';
import { useChildren } from '@vant/use';

const [name, bem] = createNamespace('row');

export type RowSpaces = { left?: number; right: number }[];

export type RowProvide = {
  spaces: ComputedRef<RowSpaces>;
};

export const ROW_KEY: InjectionKey<RowProvide> = Symbol(name);

export type RowAlign = 'top' | 'center' | 'bottom';

export type RowJustify =
  | 'start'
  | 'end'
  | 'center'
  | 'space-around'
  | 'space-between';

const rowProps = {
  tag: makeStringProp<keyof HTMLElementTagNameMap>('div'),
  wrap: truthProp,
  align: String as PropType<RowAlign>,
  gutter: makeNumericProp(0),
  justify: String as PropType<RowJustify>,
};

export type RowProps = ExtractPropTypes<typeof rowProps>;

export default defineComponent({
  name,

  props: rowProps,

  setup(props, { slots }) {
    const { children, linkChildren } = useChildren(ROW_KEY);

    const groups = computed(() => {
      const groups: number[][] = [[]];

      let totalSpan = 0;
      children.forEach((child, index) => {
        totalSpan += Number(child.span);

        if (totalSpan > 24) {
          groups.push([index]);
          totalSpan -= 24;
        } else {
          groups[groups.length - 1].push(index);
        }
      });

      return groups;
    });

    const spaces = computed(() => {
      const gutter = Number(props.gutter);
      const spaces: RowSpaces = [];

      if (!gutter) {
        return spaces;
      }

      groups.value.forEach((group) => {
        const averagePadding = (gutter * (group.length - 1)) / group.length;

        group.forEach((item, index) => {
          if (index === 0) {
            spaces.push({ right: averagePadding });
          } else {
            const left = gutter - spaces[item - 1].right;
            const right = averagePadding - left;
            spaces.push({ left, right });
          }
        });
      });

      return spaces;
    });

    linkChildren({ spaces });

    return () => {
      const { tag, wrap, align, justify } = props;
      return (
        <tag
          class={bem({
            [`align-${align}`]: align,
            [`justify-${justify}`]: justify,
            nowrap: !wrap,
          })}
        >
          {slots.default?.()}
        </tag>
      );
    };
  },
});
