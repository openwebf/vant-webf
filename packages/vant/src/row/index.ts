import { withInstall } from '../utils';
import _Row from './Row';

export const Row = withInstall(_Row);
export default Row;
export type { RowProps, RowAlign, RowJustify } from './Row';

declare module 'vue' {
  export interface GlobalComponents {
    VanRow: typeof Row;
  }
}
