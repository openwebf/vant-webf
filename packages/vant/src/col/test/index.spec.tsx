import { Col } from '..';
import { Row } from '../../row';
import { mount } from '../../../test';

test('should render Col correctly', () => {
  const wrapper = mount(Col, {
    props: {
      span: 8,
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render gutter correctly', () => {
  const wrapper = mount({
    render: () => (
      <Row gutter="24">
        <Col span="24">24</Col>

        <Col span="12">12</Col>
        <Col span="12">12</Col>

        <Col span="8">8</Col>
        <Col span="8">8</Col>
        <Col span="8">8</Col>

        <Col span="6">6</Col>
        <Col span="6">6</Col>
        <Col span="6">6</Col>
        <Col span="6">6</Col>

        <Col span="7">7</Col>
        <Col span="6">6</Col>
        <Col span="5">5</Col>
        <Col span="4">4</Col>
        <Col span="3">3</Col>
        <Col span="2">2</Col>
      </Row>
    ),
  });

  expect(wrapper.html()).toMatchSnapshot();
});
