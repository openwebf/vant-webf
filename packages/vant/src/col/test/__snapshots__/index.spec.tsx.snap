// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Col correctly 1`] = `
<div class="van-col van-col--8">
</div>
`;

exports[`should render gutter correctly 1`] = `
<div class="van-row">
  <div class="van-col van-col--24">
    24
  </div>
  <div class="van-col van-col--12">
    12
  </div>
  <div style="padding-left: 12px;"
       class="van-col van-col--12"
  >
    12
  </div>
  <div class="van-col van-col--8">
    8
  </div>
  <div style="padding-left: 12px;"
       class="van-col van-col--8"
  >
    8
  </div>
  <div style="padding-left: 16px;"
       class="van-col van-col--8"
  >
    8
  </div>
  <div class="van-col van-col--6">
    6
  </div>
  <div style="padding-left: 12px;"
       class="van-col van-col--6"
  >
    6
  </div>
  <div style="padding-left: 16px;"
       class="van-col van-col--6"
  >
    6
  </div>
  <div style="padding-left: 18px;"
       class="van-col van-col--6"
  >
    6
  </div>
  <div class="van-col van-col--7">
    7
  </div>
  <div style="padding-left: 12px;"
       class="van-col van-col--6"
  >
    6
  </div>
  <div style="padding-left: 16px;"
       class="van-col van-col--5"
  >
    5
  </div>
  <div style="padding-left: 18px;"
       class="van-col van-col--4"
  >
    4
  </div>
  <div class="van-col van-col--3">
    3
  </div>
  <div style="padding-left: 12px;"
       class="van-col van-col--2"
  >
    2
  </div>
</div>
`;
